awsmp/_driver.py

```
@@ -208,6 +227,80 @@ def _get_ratecard_info(changeset: Dict, idx: int, instance_types: List[str]) ->
+        original_pricing = {price["DimensionKey"]: price["Price"] for price in existing_prices}
+    if local_prices:
+        local_pricing = {price["DimensionKey"]: price["Price"] for price in local_prices}
+
+    diffs = []
+    for key in original_pricing:
+        if key in local_pricing and float(original_pricing[key]) != float(local_pricing[key]):
+            diffs.append(
+                {"DimensionKey": key, "Original Price": original_pricing[key], "New Price": local_pricing[key]}
+            )
+
+    return diffs
+
+
+def _check_pricing_diff(product_id: str, changeset: List[ChangeSetType]) -> Tuple[List, List]:
+    """
+    Check if there are differences between the given changeset from the local configuration and the existing listing pricing terms
+    :param str product_id: product id of existing/live listing
+    :param List[ChangeSetType] chageset: changeset from local configuration file
+    :return: Hourly and Anuual pricing diff details
+    :rtype: Tuple[List, List]
+    """
+    change = cast(dict[str, Any], changeset[0])
+    local_pricing_changesets = change["DetailsDocument"]["Terms"]
+    local_hourly, local_annual = _get_full_ratecard_info(local_pricing_changesets)
+
+    # existing pricing information from the listing
+    existing_listing_status = get_entity_details(product_id)["Description"]["Visibility"]
+    existing_terms = get_entity_details(get_public_offer_id(product_id))["Terms"]
+    existing_hourly, existing_annual = _get_full_ratecard_info(existing_terms)
+
+    if existing_listing_status != "Draft":
+        # Only staging listings are able to change pricing type
+        # e.g. free -> paid, hourly -> hourly + annual
+        if (
+            (local_hourly and not existing_hourly)
+            or (local_annual and not existing_annual)
+            or (existing_annual and not local_annual)
+        ):
+            raise Exception("Listing is published. Contact AWS Marketplace to change the pricing type")
```

we should raise a custom exception here. maybe `AMIPriceChangeError`?

---

we also need a test case that covers this branch.

---

added the custom exception for pricing type change.

also, added two more test cases to cover this change in the `tests/test_driver.py` file:
* test_get_pricing_diff_exception_no_yearly
* test_get_pricing_diff_exception_free_to_paid